main.go

```diff
@@ -72,97 +55,224 @@ func main() {
 		extra = "\n"
 	}
 
-	// output
-	fmt.Fprintf(os.Stdout, "%s%s%s", *flagPrefix, verstr, extra)
+	if *flagInverse == "" {
+		var vers []string
+		if vers, err = getVersion(wd); err != nil {
+			return err
+		}
+		if n := len(vers) - 1; *flagShort && vers[n] == "0" {
+			vers = vers[:n]
+		}
+		fmt.Fprint(os.Stdout, *flagPrefix, strings.Join(vers, *flagSep), extra)
+	} else {
+		var hash string
+		if hash, err = getInverse(wd); err != nil {
+			return err
+		}
+		fmt.Fprint(os.Stdout, hash, extra)
+	}
+
+	return nil
 }
 
 // getVersion determines the version.
-func getVersion(wd string) ([]int, error) {
+func getVersion(wd string) ([]string, error) {
+	var err error
+
+	// open repository
 	repo, err := git.PlainOpen(wd)
 	if err != nil {
 		return nil, err
 	}
-	hash, err := repo.ResolveRevision(plumbing.Revision(*flagRev))
+
+	// find commit
+	var hash *plumbing.Hash
+	var commit *object.Commit
+	if hash, err = repo.ResolveRevision(plumbing.Revision(*flagRev)); err != nil {
+		// could not resolve rev, so search for associated object
+		if h := plumbing.NewHash(*flagRev); h != plumbing.ZeroHash {
+			var obj object.Object
+			if obj, err = repo.Object(plumbing.AnyObject, h); err != nil {
+				return nil, errors.New("invalid ref")
+			}
+			var ok bool
+			if commit, ok = obj.(*object.Commit); !ok {
+				return nil, errors.New("ref is not a commit ref")
+			}
+		}
+	} else {
+		// rev flag was blank or was valid gitrev
+		commit, err = repo.CommitObject(*hash)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// empty repository
+	if commit == nil {
+		return []string{"0", "0", "0", "0"}, nil
+	}
+
+	// determine year offset
+	year, err := determineYearOffset(commit)
 	if err != nil {
 		return nil, err
 	}
 
-	// get time of commit
-	commit, err := repo.CommitObject(*hash)
+	// clamp date to UTC and year to 0
+	date := commit.Committer.When.UTC()
+	year = date.Year() - year
+	if year < 0 {
+		year = 0
+	}
+
+	// determine order
+	order, err := commitOrder(commit, date.Truncate(24*time.Hour))
 	if err != nil {
 		return nil, err
 	}
-	t := commit.Author.When.UTC()
 
-	var year int
-	// set default year based on initial commit
-	if *flagYear != "" {
-		// process year
-		if year, err = strconv.Atoi(*flagYear); err != nil {
-			return nil, err
+	return []string{
+		strconv.Itoa(year),
+		strconv.Itoa(int(date.Month())),
+		strconv.Itoa(date.Day()),
+		strconv.Itoa(order),
+	}, nil
+}
+
+// getInverse determines the hash based on the supplied flags.
+func getInverse(wd string) (string, error) {
+	var err error
+
+	// open repository
+	repo, err := git.PlainOpen(wd)
+	if err != nil {
+		return "", err
+	}
+
+	// get head
+	head, err := repo.Head()
```

Comment:
Surely this should be using flagRev?

Reply:
No -- the inverse would look at what the current head is, and walk backward. `git-buildnumber`. You wouldn't know what the `-rev` value would be, if you're inversing the version string.